docs/src/appendix/cheatcodes/l1_handler_call.md

```diff
@@ -0,0 +1,73 @@
+# `l1_handler_call`
+
+> `fn l1_handler_call(prepared_l1_handler: PreparedL1Handler)`
+
+Calls a `#[l1_handler]` function to mock a
+[message](https://docs.starknet.io/documentation/architecture_and_concepts/L1-L2_Communication/messaging-mechanism/)
+arriving from Ethereum.
+
+```rust
+struct PreparedL1Handler {
+    contract_address: ContractAddress,
+    selector: felt252,
+    from_address: felt252,
+    payload: Span::<felt252>,
+}
+```
+
+where:
+
+- `contract_address` - target contract address
+- `selector` - selector of a `#[l1_handler]` function
+- `from_address` - ethereum address which sends the message
+- `payload` - message payload
+
+It is important to note that the `from_address` must always be
+checked as any L1 contract can call any L2 contract.
+
+The payload contains any Cairo data structure that can be serialized with
+[Serde](https://book.cairo-lang.org/appendix-03-derivable-traits.html?highlight=serde#serializing-with-serde).
+
+For contract implementation:
+
+```rust
+// ...
+#[storage]
+struct Storage {
+    l1_caller: felt252,
+    //...
+}
+
+//...
+
+#[l1_handler]
+fn process_l1_message(ref self: ContractState, from_address: felt252, data: Span<felt252>) {
+    assert(from_address == self.l1_caller.read(), 'Unauthorized l1 caller');
+}
+// ...
+```
+
+We can use `l1_handler_call` to call a `l1_handler`, which is not accessible using a dispatcher:
+
+```rust
+use snforge_std::l1_handler_call;
+
+#[test]
+fn test_l1_handler_call() {
+    // ...
+    let data: Array<felt252> = array![1, 2];
+
+    let mut payload_buffer: Array<felt252> = ArrayTrait::new();
+    data.serialize(ref payload_buffer);
+
+    let l1_handler_prepared = PreparedL1Handler {
+        contract_address,
+        selector: 0x01e6b389ca484cb6fb23cbbcaa2db5581a8d970e3c135e8170c2ea5fdc2d3d8e,
```

Comment:
I think we should have similar flow here as @war-in is building for `expect-events`. @war-in can you elaborate on it here? Selector is a hashed function name so we probably can use funcionality introduced in https://github.com/foundry-rs/starknet-foundry/pull/466/files/fdd130349316dc45f3826221b034331c6dbf1e27..f386ba42c88e63f7c46bdfa86a9207e950de0f9d ?

The perfect solution would be to take selector as an argument and expose a function to calculate selector from function name.

Reply:
I would be more in favour of using raw name and calculating the entry point selector under the hood. I don't see real UX benefits coming from forcing users to manually calculate selectors

Reply:
I agree with @piotmag769. We have to use the `event_name_hash` cheatcode because the hash can't be calculated with a standard keccak syscall.
I think it does not applicate here. We can pass shortstring and calculate selector under the hood

Reply:
Agreed, important to note: this interface should probably be consistent with mock_call interface

Reply:
All right, I'll then go in that direction with the selector. :+1:

@MaksymilianDemitraszek in that case there is usually no return value. We may include it? Like this if someone wants the `l1_handler` to return a value for testing he can (which is also viable in production, just that the value is never used).

Reply:
Not sure if I understood, you would like to provide a way of asserting if certain value is returned by handler? 

I think there is no point in providing such assertion atm as it is not consumed in any way, IMHO triggering is enough

Reply:
Ok ok we've the same opinion, I was confused by "having the same interface as mock_call" sorry. Let's keep it with the invocation/trigger as it's normally done. :+1: